### DISCUSSION TITLE
Documentation suggestion

### DISCUSSION THREAD

**Discussion by JJena** [abstracta/jmeter-java-dsl#167] [General]
I've used Jmeter quite heavily. I really love this project. I want to learn more of this and use most features of this tool. Just sharing some inputs as an End user.

I couldn't find much documentation except some examples in the user guide section. For example, Jmeter has a defined structure for all of its elements like their scope, pre/post processor, child/parent. What child elements are possible for a selected element was possible in Jmeter. Right now I'm having bit difficulty to know what should go where and what dsl options are possible. Also, Apache maintains documentations for all of its Jmeter packages/classes/methods etc. If this project can create some sort of documentation in future of that sort we can get more details about the dsl specific java packages/classes etc.

**Comment by rabelenda** [comment]
Hello, sorry I missed this discussion for so long :).

First of all, thank you very much! Is really nice to get ideas for improvement and clarification, get feedback, and we really value documentation and invest on it.

Regarding what is possible to use in each part of the tree (child elements), if you use an IDE the IDE autocompletion should list at any particular context the elements that can be used at that level, for example in IntelliJ ctrl+space. You get also all the elements option with IDE auto completion when you add the dot (since we use fluent API design). Some of the videos on [this section](https://github.com/abstracta/jmeter-java-dsl#articles--talks) might help demonstrate this.

I think anyways that we could auto generate something, kind of separate from main user guide, like a components reference card. There is also the automatic javadoc generation to help with this. For instance here you can see all the elements that can me added as child of thread group [here](https://javadoc.io/static/us.abstracta.jmeter/jmeter-java-dsl/1.9/us/abstracta/jmeter/javadsl/core/threadgroups/BaseThreadGroup.ThreadGroupChild.html) checking the 
"all implementing classes" section. Or elements that can be added as [child of a sampler](https://javadoc.io/static/us.abstracta.jmeter/jmeter-java-dsl/1.9/us/abstracta/jmeter/javadsl/core/samplers/BaseSampler.SamplerChild.html). The main downside of javadocs nowadays is that it doesn't include elements in modules outside of jmeter-java-dsl. For instance, jdbc elements are not listed in implementing classes, since is part of a separate jar.

Can you clarify about this "Also, Apache maintains documentations for all of its Jmeter packages/classes/methods etc"? Maybe giving an example of how Apache does this can help.

Regards, and thank you again for giving great feedback and ideas for improvement.
